File: chirp-control-web/src/utils/websocketController.ts
export type MessageListener = (data: Record<string, unknown>) => void;
export type DisconnectListener = () => void;

export class WebSocketService {
  private deviceId: string;
  private apiUrl: string;
  private socket: WebSocket | null = null;
  private messageListeners = new Set<MessageListener>();
  private disconnectListeners = new Set<DisconnectListener>();

  constructor(
    deviceId: string,
    apiUrl = "wss://ywh1uzhhk9.execute-api.us-east-2.amazonaws.com/test",
  ) {
    this.deviceId = deviceId;
    this.apiUrl = apiUrl;
  }

  connect(): Promise<void> {
    const url = `${this.apiUrl}?deviceId=${encodeURIComponent(this.deviceId)}`;
    console.log(`Connecting to ${url} ...`);

    // Callers re-register their message/disconnect listeners after every
    // successful connect() (including reconnects) — clear stale ones first
    // so a reconnect doesn't leave the old socket's listeners stacked on
    // top of the new ones.
    this.messageListeners.clear();
    this.disconnectListeners.clear();

    return new Promise((resolve, reject) => {
      let settled = false;
      const socket = new WebSocket(url);
      this.socket = socket;

      socket.onopen = () => {
        settled = true;
        console.log("Connected to WebSocket");
        resolve();
      };

      socket.onmessage = (event) => {
        try {
          const decoded = JSON.parse(event.data);
          for (const listener of this.messageListeners) listener(decoded);
        } catch {
          console.log(`Invalid message: ${event.data}`);
        }
      };

      socket.onerror = (event) => {
        console.log("WebSocket error", event);
        if (!settled) {
          settled = true;
          reject(event);
        } else {
          for (const listener of this.disconnectListeners) listener();
        }
      };

      socket.onclose = () => {
        console.log("WebSocket closed");
        if (!settled) {
          settled = true;
          reject(new Error("WebSocket closed before opening"));
        } else {
          for (const listener of this.disconnectListeners) listener();
        }
      };
    });
  }

  onMessage(listener: MessageListener): () => void {
    this.messageListeners.add(listener);
    return () => this.messageListeners.delete(listener);
  }

  onDisconnect(listener: DisconnectListener): () => void {
    this.disconnectListeners.add(listener);
    return () => this.disconnectListeners.delete(listener);
  }

  sendCommand(command: Record<string, unknown>): void {
    if (!this.socket || this.socket.readyState !== WebSocket.OPEN) {
      console.log("WebSocket not connected.");
      return;
    }
    const payload = JSON.stringify(command);
    this.socket.send(payload);
    console.log(`Sent: ${payload}`);
  }

  disconnect(): void {
    this.messageListeners.clear();
    this.disconnectListeners.clear();
    this.socket?.close(1000);
    this.socket = null;
    console.log("Disconnected WebSocket");
  }
}

File: chirp_control/lib/utils/websocket_controller.dart
import 'dart:async';
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:web_socket_channel/status.dart' as status;

class WebSocketService {
  final String deviceId;
  final String apiUrl;

  WebSocketChannel? _channel;
  StreamSubscription? _listener;

  WebSocketService({
    required this.deviceId,
    this.apiUrl = "wss://ywh1uzhhk9.execute-api.us-east-2.amazonaws.com/test",
  });

  final _incomingController =
      StreamController<Map<String, dynamic>>.broadcast();

  Stream<Map<String, dynamic>> get messages => _incomingController.stream;

  Future<void> connect() async {
    final uri = Uri.parse("$apiUrl?deviceId=$deviceId");
    debugPrint("Connecting to $uri ...");

    try {
      _channel = WebSocketChannel.connect(uri);

      await _channel!.ready;

      _listener = _channel!.stream.listen(
        (event) {
          try {
            final decoded = jsonDecode(event);
            _incomingController.add(decoded);
          } catch (e) {
            debugPrint("Invalid message: $event");
          }
        },
        onError: (error) {
          debugPrint("WebSocket error: $error");
          throw error;
        },
        onDone: () => debugPrint("WebSocket closed"),
      );

      debugPrint("Connected to WebSocket ✅");
    } catch (e) {
      debugPrint("Connection protocol failed: $e");
      rethrow;
    }
  }

  void sendCommand(Map<String, dynamic> command) {
    if (_channel == null) {
      debugPrint("WebSocket not connected.");
      return;
    }
    final payload = jsonEncode(command);
    _channel!.sink.add(payload);
    debugPrint("Sent: $payload");
  }

  Future<void> disconnect() async {
    await _listener?.cancel();
    await _channel?.sink.close(status.normalClosure);
    debugPrint("Disconnected WebSocket");
  }
}
